cache/models/samples.js

```diff
@@ -452,6 +457,292 @@ function upsertOneSample(sampleQueryBodyObj, method, isBulk) {
+        updatedSamp, constants.fieldsToStringify.sample
+      );
+
+      // aspect is not attached to match existing behaviour
+      return sampleRes;
+    });
+  },
+
+  /**
+   * Put sample. First get sample, if not found, throw error, else get aspect.
+   * Update request body with required fields based on value and related links
+   * if needed. Delete message body and message code fields from hash
+   * if not provided because they will be null. Then update sample.
+   * @param  {Object} params - Request parameters
+   * @returns {Promise} - Resolves to a sample object
+   */
+  putSample(params) {
+    const sampleName = params.key.value;
+    const reqBody = params.queryBody.value;
+    let currSampObj;
+    let aspectObj;
+
+    return redisOps.getHashPromise(sampleType, sampleName)
+    .then((sampObj) => {
+      if (!sampObj) {
+        throw new redisErrors.ResourceNotFoundError({
+          explanation: 'Sample not found.',
+        });
+      }
+
+      currSampObj = sampObj;
+      const aspectName = sampleName.split('|')[ONE];
+      return redisOps.getHashPromise(aspectType, aspectName);
+    })
+    .then((aspObj) => {
+      if (!aspObj) {
+        throw new redisErrors.ResourceNotFoundError({
+          explanation: 'Aspect not found.',
+        });
+      }
+
+      cleanQueryBodyObj(reqBody);
+      let value = '';
+      if (reqBody.value) {
+        value = reqBody.value;
+      }
+
+      aspectObj = sampleStore.arrayStringsToJson(
+        aspObj, constants.fieldsToStringify.aspect
+      );
+
+      // change these only if status is updated
+      const status = sampleUtils.computeStatus(aspectObj, value);
+      if (currSampObj[sampFields.STATUS] !== status) {
+        reqBody[sampFields.PRVS_STATUS] = currSampObj[sampFields.STATUS];
+        reqBody[sampFields.STS_CHNGED_AT] = new Date().toString();
+        reqBody[sampFields.STATUS] = status;
+      }
+
+      // We have a value, so update value and updated_at always.
+      reqBody[sampFields.VALUE] = value;
+      reqBody[sampFields.UPD_AT] = new Date().toString();
+
+      if (reqBody.relatedLinks) { // related links
+        reqBody[sampFields.RLINKS] = JSON.stringify(reqBody.relatedLinks);
+      } else {
+        reqBody[sampFields.RLINKS] = '[]';
+      }
+
+      const cmds = [];
+
+      // delete fields with future null value
+      if (!reqBody.messageBody) {
+        cmds.push(
+          redisOps.delHashFieldCmd(sampleType, sampleName, sampFields.MSG_BODY)
+        );
+      }
+
+      if (!reqBody.messageCode) {
+        cmds.push(
+          redisOps.delHashFieldCmd(sampleType, sampleName, sampFields.MSG_CODE)
+        );
+      }
+
+      cmds.push(redisOps.setHashMultiCmd(sampleType, sampleName, reqBody));
+      return redisOps.executeBatchCmds(cmds);
+    })
+    .then(() => redisOps.getHashPromise(sampleType, sampleName))
+    .then((updatedSamp) => cleanAddAspectToSample(updatedSamp, aspectObj));
+  },
+
+  /**
```

Comment:
The upsert endpoints are also passed through the responsify method before returning the response.